steering_controllers_library/src/steering_controllers_library.cpp

```diff
@@ -537,12 +603,23 @@ controller_interface::CallbackReturn SteeringControllersLibrary::on_deactivate(
 controller_interface::return_type SteeringControllersLibrary::update_reference_from_subscribers(
   const rclcpp::Time & /*time*/, const rclcpp::Duration & /*period*/)
 {
-  auto current_ref = *(input_ref_.readFromRT());
-
-  if (!std::isnan(current_ref->twist.linear.x) && !std::isnan(current_ref->twist.angular.z))
+  if (params_.twist_input)
+  {
+    auto current_ref = *(input_ref_twist_.readFromRT());
+    if (!std::isnan(current_ref->twist.linear.x) && !std::isnan(current_ref->twist.angular.z))
+    {
+      reference_interfaces_[0] = current_ref->twist.linear.x;
+      reference_interfaces_[1] = current_ref->twist.angular.z;
+    }
+  }
+  else
   {
-    reference_interfaces_[0] = current_ref->twist.linear.x;
-    reference_interfaces_[1] = current_ref->twist.angular.z;
+    auto current_ref = *(input_ref_ackermann_.readFromRT());
+    if (!std::isnan(current_ref->drive.speed) && !std::isnan(current_ref->drive.steering_angle))
+    {
+      reference_interfaces_[0] = current_ref->drive.speed;
+      reference_interfaces_[1] = current_ref->drive.steering_angle;
```

Comment:
```suggestion
      reference_interfaces_[0] = current_ref->drive.speed;
      reference_interfaces_[1] = current_ref->drive.steering_angle;
```

I have a question: If we only use speed and steering_angle, is it a good idea to use the full blown [ackermann_msg](https://github.com/ros-drivers/ackermann_msgs/blob/ros2/msg/AckermannDrive.msg) definition? Furthermore, I am reluctant using a message type called ackermann_msg, even if the robot has no ackermann kinematics. Is there a reason for that, for example an upstream package which publishes in this format? Otherwise, I'd suggest to add a new message similar to [this](https://github.com/ros-controls/control_msgs/blob/master/control_msgs/msg/SteeringControllerStatus.msg) to control_msgs repository.

Reply:
I mainly used this message since it was already present prior to the removal of the message publisher of this type in the initial draft of this PR. I initially also thought that it may be good to have some cross-compatibility, but I am not aware of big projects that use this message atm. So I don't have any hard feelings on using a custom message for this. How would you call it instead?

Reply:
> How would you call it instead?

A good question :grin: Maybe simply `SteeringControllerCommand`? In accordance with https://github.com/ros-controls/control_msgs/blob/master/control_msgs/msg/SteeringControllerStatus.msg

Reply:
Good idea, I created a PR: https://github.com/ros-controls/control_msgs/pull/217